# 01. Importing libraries

In [4]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [5]:
# Importing datasets

path = r'C:\Users\Amy Yip\Desktop\Career Foundry Data Analytics\Achievement 4 Python\05-02-2022 Instacart Basket Analysis'

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 02. Perform wrangling procedures walked through from Exercise Learning Notes

* Dropping columns
* Renaming columns
* Changing data types
* Transposing data

## (i) Dropping Column

In [6]:
# Drop eval_set column from orders.csv. Following shows the results visually only:

df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
5         3367565        1             6          2                  7   
6          550135        1             7          1                  9   
7         3108588        1             8          1                 14   
8         2295261        1             9          1                 16   
9         2550362        1            10          4                  8   
10        1187899        1            11          4                  8   
11        2168274        2             1          2                 11   
12        1501582        2             2          5                 10   
13        1901567        2             3          1                 10   
14         738281        2             4          2                 10   
15        1673511        2             5          3                 11   
16        1199898        2             6          2                  9   
17        3194192        2             7          2                 12   
18         788338        2             8          1                 15   
19        1718559        2             9          2                  9   
20        1447487        2            10          1                 11   
21        1402090        2            11          1                 10   
22        3186735        2            12          1                  9   
23        3268552        2            13          4                 11   
24         839880        2            14          3                 10   
25        1492625        2            15          1                 11   
26        1374495        3             1          1                 14   
27         444309        3             2          3                 19   
28        3002854        3             3          3                 16   
29        2037211        3             4          2                 18   
...           ...      ...           ...        ...                ...   
3421053   2789700   206208            35          3                 22   
3421054    844592   206208            36          6                 15   
3421055   1541132   206208            37          2                 12   
3421056   2808240   206208            38          0                 15   
3421057   3027766   206208            39          2                 14   
3421058   3356245   206208            40          5                  9   
3421059    442304   206208            41          2                 14   
3421060   2675140   206208            42          1                 19   
3421061    167903   206208            43          4                 14   
3421062   2393201   206208            44          6                 16   
3421063   3292671   206208            45          2                 11   
3421064   3059777   206208            46          1                 10   
3421065   2239861   206208            47          3                  4   
3421066   1285346   206208            48          1                 11   
3421067   1882108   206208            49          1                 22   
3421068    803273   206208            50          5                 11   
3421069   3154581   206209             1          3                 11   
3421070   1889163   206209             2          3                 17   
3421071   1542354   206209             3          5                 11   
3421072    688306   206209             4          1                 10   
3421073   2307371   206209             5          4                 15   
3421074   3186442   206209             6          0                 16   
3421075    55

In [7]:
# Drop eval_set column from orders.csv. Following shows the results with the column actually removed:

df_ords = df_ords.drop(columns = ['eval_set'])

## Missing Data

In [8]:
# Additional from the course notes
# Check if there are missing value in "days_since_prior_order" columns:

df_ords['days_since_prior_order'].value_counts(dropna = False)

 30.0    369323
 7.0     320608
 6.0     240013
 4.0     221696
 3.0     217005
 5.0     214503
NaN      206209
 2.0     193206
 8.0     181717
 1.0     145247
 9.0     118188
 14.0    100230
 10.0     95186
 13.0     83214
 11.0     80970
 12.0     76146
 0.0      67755
 15.0     66579
 16.0     46941
 21.0     45470
 17.0     39245
 20.0     38527
 18.0     35881
 19.0     34384
 22.0     32012
 28.0     26777
 23.0     23885
 27.0     22013
 24.0     20712
 25.0     19234
 29.0     19191
 26.0     19016
Name: days_since_prior_order, dtype: int64

## (ii) Renaming Column

In [9]:
# Rename "order_dow" column:

df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [10]:
# Check whether the code was executed successfully:

df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## (iii) Changing Data Type

In [11]:
# Change data type of "order_id" column to a string:

df_ords['order_id'] = df_ords['order_id'].astype('str')

In [12]:
# Return data type of new "order_id" column:

df_ords['order_id'].dtype

dtype('O')

# (iv) Transposing Data

In [13]:
# Importing departments.csv data

df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [14]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9   ...               12      13         14  \
0  beverages  pets  dry goods pasta   ...     meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [15]:
# Transposing df_dep

df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [16]:
# Overwrite old df_dep dataframe with new transposed df_dep dataframe 

df_dep_t = df_dep.T

In [17]:
# Check new dataframe

df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [18]:
# Add an index to the dataframe 

df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [19]:
# Take the first row of df_dep_t as the header

# Note: iloc() function enables us to select a particular cell of the dataset.

new_header = df_dep_t.iloc[0]

In [20]:
new_header

0    department
Name: department_id, dtype: object

In [21]:
# Create a new dataframe that only copies over rows beyond the first row. 

df_dep_t_new = df_dep_t[1:]

In [22]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [23]:
# Set the header row as the df header.

df_dep_t_new.columns = new_header

In [24]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# 03. Data Dictionary

In [25]:
# Turn department dataframe into dictionary

data_dict = df_dep_t_new.to_dict('index')

In [26]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [27]:
# Using data dictionary to find what "department_id" of 19 corresponds to within the dataframe.

print(data_dict.get('19'))


{'department': 'snacks'}


# 04. Subsetting

In [28]:
# Creating a subset for products dataframe to contain only data from the snacks department

df_prods['department_id'] == 19

0         True
1        False
2        False
3        False
4        False
5        False
6        False
7        False
8        False
9        False
10       False
11       False
12       False
13       False
14       False
15        True
16       False
17       False
18       False
19       False
20       False
21       False
22       False
23       False
24        True
25       False
26       False
27       False
28       False
29       False
         ...  
49663    False
49664    False
49665    False
49666     True
49667    False
49668    False
49669     True
49670     True
49671    False
49672    False
49673    False
49674    False
49675     True
49676    False
49677    False
49678    False
49679    False
49680    False
49681    False
49682    False
49683     True
49684    False
49685    False
49686    False
49687    False
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [29]:
df_prods[df_prods['department_id'] == 19]

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
15             16                      Mint Chocolate Flavored Syrup   
24             25            Salted Caramel Lean Protein & Fiber Bar   
31             32                      Nacho Cheese White Bean Chips   
40             41        Organic Sourdough Einkorn Crackers Rosemary   
55             56                         Healthy Pop Butter Popcorn   
58             59               Medium Taqueria Style Chipotle Salsa   
76             77                  Coconut Chocolate Chip Energy Bar   
77             78                  Nutter Butter Cookie Bites Go-Pak   
92             93                        Uncured Cracked Pepper Beef   
101           102                              Danish Butter Cookies   
102           103                              Sprouted Kale Cracker   
114           115                  Scooby-Doo! Fruit Flavored Snacks   
123           124                            Sun Cups Dark Chocolate   
134           135                   Dark Chocolate Ice Cream Topping   
135           136                                             Simple   
139           140                        Chips Onion Chipotle Garlic   
140           141  Restaurant Style Organic Chia & Quinoa Tortill...   
144           145                         Traditional Water Crackers   
155           156                     Extra Spearmint Sugar-Free Gum   
163           164                Peanut Butter Cacao Quinoa Clusters   
165           166                   Garlic Parmesan Pita Bread Chips   
171           172     Gluten Free All Natural Chocolate Chip Cookies   
179           180            Simply Beyond Black Bean Tortilla Chips   
182           183                    Cookie Dough Performance Energy   
183           184                        Kisses Milk Chocolate Candy   
204           205         Chocolate Mint Cookie Crunch Nutrition Bar   
206           207                                   Minis Candy Bars   
211           212          Organic Blue Corn Sea Salt Tortilla Chips   
212           213     Tapatio Salsa Picante Hot Sauce Tortilla Chips   
...           ...                                                ...   
49456       49452                          Garlic Herb Pumpkin Seeds   
49479       49475                    Dark Chocolate Blueberry Quinoa   
49496       49492                                  Gummies 5 Flavors   
49497       49493                             75% Dark Chocolate Bar   
49500       49496  Puffs Simply White Cheddar Cheese Flavored Snacks   
49508       49504       Dark Chocolate Peanut Butter Cups Miniatures   
49514       49510                   Lentil Chips, Aged White Cheddar   
49516       49512                                Tiny Twists Prezels   
49518       49514  Baked Triscuits with Whole Grain Wheat Mediter...   
49526       49522                            Cashews Halves & Pieces   
49529       49525                   Cinnamon Roll Flavor Protein Bar   
49532       49528                            Sweet Onion Wheat Thins   
49538       49534                     Jalapeño Seasoned Potato Chips   
49565       49561                         Crunchy Peanut Butter Bars   
49568       49564                         Hot Dark Chocolate Topping   
49570       49566  Peanut Butter Dark Chocolate Fruit & Nut Prote...   
49581       49577                            Wavy Ranch Potato Chips   
49595       49591                               Fruit Roll Raspberry   
49597       49593                      Classic Movie Theater Popcorn   
49598       49594                                     Yellow Edition   
49605       49601                            Pomegranate Gummy Bears   
49618       49614                   Sandies Pecan Shortbread Cookies   
49622       49618                                Mac n' Cheese Puffs   
49624       49620                              

In [30]:
df_snacks =  df_prods[df_prods['department_id']==19]

In [31]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [ ]:
# Another option is using loc function, which looks for value of "department_id" is 19:

df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

In [ ]:
# Final option to use is:

df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

# 04. Data Wrangling Procedures for the Ex 4.4 Steps

2. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [32]:
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0        2539329        1             1                   2   
1        2398795        1             2                   3   
2         473747        1             3                   3   
3        2254736        1             4                   4   
4         431534        1             5                   4   
5        3367565        1             6                   2   
6         550135        1             7                   1   
7        3108588        1             8                   1   
8        2295261        1             9                   1   
9        2550362        1            10                   4   
10       1187899        1            11                   4   
11       2168274        2             1                   2   
12       1501582        2             2                   5   
13       1901567        2             3                   1   
14        738281        2             4                   2   
15       1673511        2             5                   3   
16       1199898        2             6                   2   
17       3194192        2             7                   2   
18        788338        2             8                   1   
19       1718559        2             9                   2   
20       1447487        2            10                   1   
21       1402090        2            11                   1   
22       3186735        2            12                   1   
23       3268552        2            13                   4   
24        839880        2            14                   3   
25       1492625        2            15                   1   
26       1374495        3             1                   1   
27        444309        3             2                   3   
28       3002854        3             3                   3   
29       2037211        3             4                   2   
...          ...      ...           ...                 ...   
3421053  2789700   206208            35                   3   
3421054   844592   206208            36                   6   
3421055  1541132   206208            37                   2   
3421056  2808240   206208            38                   0   
3421057  3027766   206208            39                   2   
3421058  3356245   206208            40                   5   
3421059   442304   206208            41                   2   
3421060  2675140   206208            42                   1   
3421061   167903   206208            43                   4   
3421062  2393201   206208            44                   6   
3421063  3292671   206208            45                   2   
3421064  3059777   206208            46                   1   
3421065  2239861   206208            47                   3   
3421066  1285346   206208            48                   1   
3421067  1882108   206208            49                   1   
3421068   803273   206208            50                   5   
3421069  3154581   206209             1                   3   
3421070  1889163   206209             2                   3   
3421071  1542354   206209             3                   5   
3421072   688306   206209             4                   1   
3421073  2307371   206209             5                   4   
3421074  3186442   206209             6                   0   
3421075   550836   206209             7                   2   
3421076  2129269   206209             8                   3   
3421077  2558525   206209             9                   4   
3421078  2266710   206209            10                   5   
3421079  1854736   206209            11                   4   
3421080   626363   206209            12                   1   
3421081  2977660   206209            13                   1   
3421082   272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     N

From course notes: 
There would never be a need to calculate statistics for the "order_id" and "user_id" columns as they don’t represent real numeric values, rather, keys related to specific orders and users.

These 2 columns are known as key characteristics because they act as keys for merging with other data.

In [33]:
# Since order_id variable data type has already been changed in previous steps,therefore,
# here, change user_id variable (numeric) to string.

df_ords['user_id'] = df_ords['user_id'].astype('str')

In [34]:
df_ords['user_id'].dtype

dtype('O')

3. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the data frame.

In [35]:
# “order_number” column represents how many orders this particular user has made.

# Rename “order_number” to a more intuitive name order_numbers_users_made

df_ords.rename(columns = {'order_number':'order_numbers_users_made'}, inplace = True)

In [36]:
df_ords.head()

order_id user_id  order_numbers_users_made  orders_day_of_week  \
0  2539329       1                         1                   2   
1  2398795       1                         2                   3   
2   473747       1                         3                   3   
3  2254736       1                         4                   4   
4   431534       1                         5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

### df.value_counts --> definition: return counts of rows

#### df.value_counts(subset=None, normalize=False, sort=True, ascending=False, dropna=True)


subset: list-like, optional
Columns to use when counting unique combinations.

normalize: bool, default False
Return proportions rather than frequencies.

sort: bool, default True
Sort by frequencies.

ascending: bool, default False
Sort in ascending order.

dropna: bool, default True
Don’t include counts of rows that contain NA values.

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.value_counts.html

In [37]:
df_ords['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

The busiest hour of the day for placing orders is at 10am with a total of 288418 orders.

5. Determine the meaning behind a value of 4 in the "department_id" column within the dataframe using a data dictionary.

In [38]:
print(data_dict.get('4'))

{'department': 'produce'}


The department_id 4 is the produce department.

6. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [39]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

Try nested loops to find the department_id of 'breakfast' in data_dict

https://stackoverflow.com/questions/10756427/loop-through-all-nested-dictionary-values

In [53]:
print(data_dict.get('breakfast'))

None


Looking at the data_dict above, breakfast department corresponds to department_id = 14

In [42]:
# Creating a subset of breakfast item sales from products dataframe:

df_breakfast = df_prods[df_prods['department_id'] == 14]

In [43]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

7. They’d also like to see details about customers who might be throwing dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [44]:
# Creating a subset containing items from departments; alcohol(5), deli(20), beverages(7),
# and meat/seafood(12)

df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5, 20, 7, 12])]

In [45]:
df_dinner_parties.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

8. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [46]:
df_dinner_parties.shape

(7650, 5)

9. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [47]:
df_user_1 = df_ords[df_ords['user_id']=='1']

# Note: user_id column was converted into a string earlier, therefore need to use quotation marks

In [48]:
df_user_1

order_id user_id  order_numbers_users_made  orders_day_of_week  \
0   2539329       1                         1                   2   
1   2398795       1                         2                   3   
2    473747       1                         3                   3   
3   2254736       1                         4                   4   
4    431534       1                         5                   4   
5   3367565       1                         6                   2   
6    550135       1                         7                   1   
7   3108588       1                         8                   1   
8   2295261       1                         9                   1   
9   2550362       1                        10                   4   
10  1187899       1                        11                   4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

10. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [49]:
df_user_1.describe()

order_numbers_users_made  orders_day_of_week  order_hour_of_day  \
count                 11.000000           11.000000          11.000000   
mean                   6.000000            2.636364          10.090909   
std                    3.316625            1.286291           3.477198   
min                    1.000000            1.000000           7.000000   
25%                    3.500000            1.500000           7.500000   
50%                    6.000000            3.000000           8.000000   
75%                    8.500000            4.000000          13.000000   
max                   11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

11. Check the organization and structure of your notebook. Be sure to include section headings and code comments.

12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [50]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled_ex4.4_task.csv'))

13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [51]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled_ex4.4_task.csv'))

14. Save your Jupyter notebook.